pkg/kubelet: add KUBE-FIREWALL-IN chain, move localnet rule to it

squeed opened the pull request:
@champtar observed that the rule added in #91569 still allowed packets through unidirectionally, meaning UDP with a spoofed source wouldn't be caught. So we need to filter on source interface.

However, KUBE-FIREWALL is run on both OUTPUT and INPUT, so it can't reference interfaces. This PR, thus,
- creates a chain KUBE-FIREWALL-IN
- Add it to INPUT, pass through to KUBE-FIREWALL
- Move the localnet protection rule to KUBE-FIREWALL-IN, and adjust it
  to reference source if, not source address.

Signed-off-by: Casey Callendrello <cdc@redhat.com>


**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #90259

**Special notes for your reviewer**:
You can test this pretty simply using `cnitool` if you don't want to spin up a whole kind cluster.


**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

/sig network

squeed commented:
/cc @derekwaynecarr @matthyx

squeed commented:
/hold
@champtar we can't do this - KUBE-FIREWALL is called in both INPUT and OUTPUT, and you can't do `-i foo` in an OUTPUT rule.

champtar commented:
@squeed good catch, I only tested that `127.0.0.1:NodePort` was working, nothing more ...
In that case you need a KUBE-FIREWALL-IN and KUBE-FIREWALL-OUT that chain to KUBE-FIREWALL

squeed commented:
Created KUBE-FIREWALL-IN, but not KUBE-FIREWALL-OUT, since we don't need it (yet). If this should exist for symmetry, I'm happy to add it.

squeed commented:
/hold cancel

squeed commented:
/cc @danwinship

joelsmith commented:
/retest

squeed commented:
> would it be better to just create a new chain?

Do we have any sort of "API guarantees" around the existence of certain chains? Are users expecting to be able to add rules to KUBE-FIREWALL?

If not, then we should just move to KUBE-FIREWALL-IN and KUBE-FIREWALL-OUT and be done with it.

champtar commented:
@danwinship the nesting is not mandatory, we can add both to INPUT
```
INPUT
-j KUBE-FIREWALL
-j KUBE-FIREWALL-IN
```

danwinship commented:
Our iptables chains are not _supposed_ to be API, but we don't know whether people depend on them. But I meant something like what @champtar said, though with a different name than `KUBE-FIREWALL-IN`

squeed commented:
I specifically named it KUBE-FIREWALL-IN to indicate it's only for INPUT and thus can predicate on incoming interface `-i lo`. Naming it `KUBE-SECURITY` misses this context.

Annoyingly, iptables happily lets you shoot yourself in the foot; you cant do ` iptables -A OUTPUT -i lo ...`, but you can jump to a chain from OUTPUT that does.

Given that this is getting a bit uglier, I can also try and move this to the proxy code (iptables & ipvs), since they're the ones setting the offending sysctl anyways.

matthyx commented:
/uncc

squeed commented:
Hey @champtar, I'm not sure this is necessary. The packets you observed (spoofed source IP of 127.0.0.2) will be blocked by the reverse path filter. Unless, of course, that's also disabled.

Given that disabling rpf is not so common on kubernetes nodes, I think we can just close this.

champtar commented:
@squeed, ubuntu 20.04 seems to defaults to loose rp filtering, so I would say need testing
```
root@etienne-ubuntu2004:~# grep -r rp_ /etc/sysctl.d/
/etc/sysctl.d/10-network-security.conf:net.ipv4.conf.default.rp_filter=2
/etc/sysctl.d/10-network-security.conf:net.ipv4.conf.all.rp_filter=2
```

champtar commented:
Until either all CNIs or the kubelet or the kube-proxy enforce `rp_filter=1`, you should consider that it is not the case

champtar commented:
I did some more testing, and using Linux 5.6.12-300.fc32.x86_64, `rp_filter=2` drops packets with ip source in 127.0.0.0/8
So @squeed you are right it seems we need `rp_filter=0` to be vulnerable

rtreffer commented:
Hi,

The original iptables rule caused our cluster tests to fail on `kubectl exec`. The kubelet connected to the local privileged port with the node ip as the source, thus matching the conditions (`src != 127.0.0.1/8, dst = 127.0.0.1/8`)

Errors looked like this
```
# kubectl --namespace k8s exec k8s-dns-d584d5d8c-tk5bt -- echo it works
error: unable to upgrade connection: error dialing backend: dial tcp 127.0.0.1:36827: connect: connection timed out
```

After removing the firewall from OUTPUT :see_no_evil: and changing the rule I can exec again into pods
```
# iptables -R KUBE-FIREWALL 2 -m comment --comment "block incoming localnet connections" --dst 127.0.0.1/8 \! -i lo -m conntrack \! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
# iptables -F OUTPUT
```
```
kubectl --namespace k8s exec k8s-dns-d584d5d8c-tk5bt -- echo it works
it works
```

This is currently blocking recent kubernetes version for us. Please cherry-pick onto 1.18 once merged.

danwinship commented:
> I did some more testing, and using Linux 5.6.12-300.fc32.x86_64, `rp_filter=2` drops packets with ip source in 127.0.0.0/8
> So @squeed you are right it seems we need `rp_filter=0` to be vulnerable

So @champtar does that mean this PR can be closed?

champtar commented:
This PR is pretty simple and guarantee we are safe, in the current state we are not safe under some conditions.
I agree that having `rp_filter=0` is not common, but why not just be secure 100% of the time ?
Also I haven't tested older kernel.
You can do kernel code review to understand the behavior, if it changed recently, and update the doc to warn about not using `rp_filter=0`, or just fix this problem (that k8s/kubeproxy creates) once and for all with 25 lines of change.

ehashman commented:
@squeed can you rebase or should this be closed?

squeed commented:
@ehashman sure, I can pick this back up (probably not this week, but soon).

squeed commented:
/remove-lifecycle rotten

I'm picking this back up, as I have heard some indications that the simpler fix is causing some unexpected traffic blocking.

squeed commented:
Okay, this is rebased. @champtar @tssurya would you mind reviewing?

I've heard some rumbling that the current behavior blocks some legitimate traffic, but I've not been able to get a concrete example.

squeed commented:
/cc @tssurya

champtar commented on pkg/kubelet/kubelet_network_linux.go:
```diff
@@ -95,27 +97,37 @@ func (kl *Kubelet) syncNetworkUtil(iptClient utiliptables.Interface) bool {
-			"!", "--src", "127.0.0.0/8",
+			"!", "-i", "lo",
 			"-m", "conntrack",
 			"!", "--ctstate", "RELATED,ESTABLISHED,DNAT",
```
Do we still need the ctstate match ?

champtar replied:
(still before coffee) yes

squeed replied:
I believe so, yes. imagine you have

```
ipables -t nat -A PREROUTING -m tcp --dst-port 443 -j DNAT --to-destination 127.0.0.1:6443
```

while this is slightly convoluted, we should definitely allow this to pass through.

champtar commented on pkg/kubelet/kubelet_network_linux.go:
```diff
@@ -95,27 +97,37 @@ func (kl *Kubelet) syncNetworkUtil(iptClient utiliptables.Interface) bool {
+		return false
+	}
+	if _, err := iptClient.EnsureRule(utiliptables.Prepend, utiliptables.TableFilter, utiliptables.ChainInput, "-j", string(KubeFirewallInChain)); err != nil {
+		klog.Errorf("Failed to ensure that %s chain %s jumps to %s: %v", utiliptables.TableFilter, utiliptables.ChainInput, KubeFirewallInChain, err)
```
If I read correctly
```
- iptables -t filter -I INPUT -j KUBE-FIREWALL
+ iptables -t filter -A KUBE-FIREWALL-IN -j KUBE-FIREWALL
+ iptables -t filter -I INPUT -j KUBE-FIREWALL-IN
```
This works for me.
I'm curious why you didn't went with
```
  iptables -t filter -I INPUT -j KUBE-FIREWALL
+ iptables -t filter -I INPUT -j KUBE-FIREWALL-IN
```
(performance ? usability ?)

champtar replied:
Also what happens on upgrade ?

squeed replied:
Oh, upgrade! Good catch.

It was a while ago, but I was thinking that one would theoretically like to add `-j ALLOW` rules in `KUBE-FIREWALL-IN`, but that doesn't actually make any sense. I'll do what you suggest.

squeed replied:
updated.

squeed commented:
Okay, updated to remove the goofy jump rules. Now the net difference is a lot simpler:

```
+ -A INPUT -j KUBE-FIREWALL-IN
+ -A KUBE-FIREWALL-IN -d 127.0.0.0/8 ! -i lo -m comment --comment "block incoming localnet connections" -m conntrack ! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
```

The only thing this PR doesn't do is delete the old rule; That's fine, it will go away on next reboot, and it isn't hurting anything.

squeed committed:
pkg/kubelet: add KUBE-FIREWALL-IN chain, move localnet rule to it

champtar observed that the rule added in #91569 still allowed packets
through unidirectionally, meaning UDP with a spoofed source wouldn't be
caught. So we need to filter on source interface.

However, KUBE-FIREWALL is run on both OUTPUT and INPUT, so it can't
reference interfaces. This PR, thus,

- creates a chain KUBE-FIREWALL-IN
- Add it to INPUT, pass through to KUBE-FIREWALL
- Move the localnet protection rule to KUBE-FIREWALL-IN, and adjust it
  to reference source if, not source address.

Signed-off-by: Casey Callendrello <cdc@redhat.com>

champtar commented:
> Okay, updated to remove the goofy jump rules. Now the net difference is a lot simpler:
> 
> ```
> + -A INPUT -j KUBE-FIREWALL-IN
> + -A KUBE-FIREWALL-IN -d 127.0.0.0/8 ! -i lo -m comment --comment "block incoming localnet connections" -m conntrack ! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
> ```

To be exact
```
+ -I INPUT -j KUBE-FIREWALL-IN
+ -A KUBE-FIREWALL-IN -d 127.0.0.0/8 ! -i lo -m comment --comment "block incoming localnet connections" -m conntrack ! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
```
which is what we want

> The only thing this PR doesn't do is delete the old rule; That's fine, it will go away on next reboot, and it isn't hurting anything.

agreed

ship it !

tssurya commented:
> Okay, updated to remove the goofy jump rules. Now the net difference is a lot simpler:
> 
> ```
> + -A INPUT -j KUBE-FIREWALL-IN
> + -A KUBE-FIREWALL-IN -d 127.0.0.0/8 ! -i lo -m comment --comment "block incoming localnet connections" -m conntrack ! --ctstate RELATED,ESTABLISHED,DNAT -j DROP
> ```

Interesting issue, took me a while to read and catch up on the history. Checked this on KIND and it looks exactly like what you pasted ^. Tested it and it seems fine to me. Not sure if its easy to just add a simple unit test to keep track of the rule changes, of course no need to block the merge for that.

> 
> The only thing this PR doesn't do is delete the old rule; That's fine, it will go away on next reboot, and it isn't hurting anything.

okay so since we are not deleting/changing jump paths, we don't have to worry about the upgrade use case as well.

ehashman commented:
@kubernetes/sig-network-pr-reviews anyone wanna TAL at this? :)

ehashman commented:
/priority important-longterm

SergeyKanzhelev commented:
needs rebase

squeed commented:
rebased.

ehashman commented:
/assign @danwinship

ehashman commented:
also
/cc @aojea 

would love to get this one closed out @kubernetes/sig-network-bugs

aojea commented:
/cc

aojea commented:
I'm puzzled with this issue, the description of the original issue refers that traffic external to the node can send packets to the localhost of another node, and since it has route_localnet enables, external nodes should be able to reach them

> A host on the local segment can actually route 127.0.0.1 to a k8s-node and access ports that really should be local.

However, we are only filtering here traffic from/to the same node to localhost, what part I'm missing :sweat_smile: ?
